keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,999 @@
+## Design Details
+
+### ClusterNetworkPolicy API Design
+The following new `ClusterNetworkPolicy` API will be added to the `networking.k8s.io` API group:
+
+```golang
+type ClusterNetworkPolicy struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec ClusterNetworkPolicySpec
+}
+
+type ClusterNetworkPolicySpec struct {
+	// No implicit isolation of AppliedTo Pods.
+	AppliedTo    AppliedTo
+	Ingress      []ClusterNetworkPolicyIngressRule
+	Egress       []ClusterNetworkPolicyEgressRule
+}
+
+type ClusterNetworkPolicyIngress/EgressRule struct {
+	Action       RuleAction
+	Ports        []networkingv1.NetworkPolicyPort
+	From/To      []ClusterNetworkPolicyPeer
+}
+
+type ClusterNetworkPolicyPeer struct {
+	PodSelector  *metav1.LabelSelector
+	// required if a PodSelector is specified
+	Namespaces   *Namespaces
+	IPBlock      *networkingv1.IPBlock
+}
+
+const (
+  RuleActionAuthorize RuleAction = "Authorize"
+	RuleActionDeny      RuleAction = "Deny"
+	RuleActionAllow     RuleAction = "Allow"
+)
+```
+
+For the ClusterNetworkPolicy ingress/egress rule, the `Action` field dictates whether
+traffic should be allowed or denied or always allowed from/to the ClusterNetworkPolicyPeer.
+This will be a required field.
+
+### ClusterDefaultNetworkPolicy API Design
+The following new `ClusterDefaultNetworkPolicy` API will be added to the `networking.k8s.io` API group:
+
+```golang
+type ClusterDefaultNetworkPolicy struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec ClusterDefaultNetworkPolicySpec
+}
+
+type ClusterDefaultNetworkPolicySpec struct {
+	// Implicit isolation of AppliedTo Pods.
+	AppliedTo   AppliedTo
+	Ingress     []ClusterDefaultNetworkPolicyIngressRule
+	Egress      []ClusterDefaultNetworkPolicyEgressRule
+}
+
+type ClusterDefaultNetworkPolicyIngress/EgressRule struct {
+	Ports            []networkingv1.NetworkPolicyPort
+	OnlyFrom/OnlyTo  []ClusterDefaultNetworkPolicyPeer
+}
+
+type ClusterDefaultNetworkPolicyPeer struct {
+	PodSelector  *metav1.LabelSelector
+	// required if a PodSelector is specified
+	Namespaces   *Namespaces
+	IPBlock      *networkingv1.IPBlock
+}
+```
+
+Most structs above are very similar to NetworkPolicy and quite self-explanatory.
+One detail to notice is that in the ClusterDefaultNetworkPolicy Ingress/Egress rule spec,
+the peers are created in a field named `OnlyFrom`/`OnlyTo`, as opposed to `To`/`From`
+in ClusterNetworkPolicy. We chose this naming to better hint policy writers about
+the isolation effect of ClusterDefaultNetworkPolicy on the workloads it applies to.
+
+### Shared API Design
+The following structs will be added to the `networking.k8s.io` API group and
+shared between `ClusterNetworkPolicy` and `ClusterDefaultNetworkPolicy`:
+
+```golang
+type AppliedTo struct {
+	// required if a PodSelector is specified
+	NamespaceSelector   *metav1.LabelSelector
+	// optional
+	PodSelector         *metav1.LabelSelector
+}
+
+type Namespaces struct {
+	// Self and Selector are mutually-exclusive
+	Self       bool
```

Comment:
Generally nice to avoid using bool fields in favor of enums. It makes it easier to extend the field's behavior in the future.

e.g.,

```go
type Namespaces struct {
  Match NamespaceMatchType
}

type NamespaceMatchType string

const (
  NamespaceMatchSelf NamespaceMatchType = "Self"
  
  // Contrived example of something that might be added in the future.
  NamespaceMatchNotSelf NamespaceMatchtype = "NotSelf"
)
```

Reply:
We considered this and decided to go with the `bool` field because as of now there aren't other `NamespaceMatchtype` that makes sense. For the `NamespaceMatchNotSelf`, are you suggesting its meaning as "select from all namespaces except from self Namespaces"? That will be an interesting semanitcs. 
Also as @tnqn pointed out in https://github.com/kubernetes/enhancements/pull/2522#discussion_r614727344 how namespace matching and namespaceSelector co-exist with each other is another thing to consider

Reply:
> Self, Selector, SameLabels, DifferentLabels - I can name 4 off the top of my head :)

Well the problem is the other `NamespaceMatchtypes` you suggested here, user will need to provide a kv pair to argument that `NamespaceMatchtype`, correct? Whereas for `self`, this isn't necessary and ideally this should be as simple as `Namespaces: self`.
This is what we were stuggling with before deciding to go with the bool field: how those special cases/keywords can co-exist with the current MatchLabels/MatchExpression structs.

Reply:
Yes, some of those would take params, and that is OK.  e.g.

```
namespaces:
  scope: Self
```

vs

```
namespaces:
  scope: Selector
  selector:
  - foo: bar
```

vs

```
namespaces:
  scope: SameLabels
  labels:
  - mycompany.com/tenant
```